AIP-7339 Recurring Workflow

talebzeghmi opened the pull request:
- [x] synchronization and ConfigMap creation for `kfp run`

talebzeghmi committed:
AIP-7339 first checkin

cloudw commented on metaflow/plugins/kfp/kfp.py:
```diff
@@ -186,7 +187,7 @@ def __init__(
         try:
             # Register workflow template.
-            workflow: Dict[Text, Any] = self._create_workflow_yaml(
+            workflow = self._create_workflow_yaml(
```
Do you prefer not having the type hint here?

talebzeghmi replied:
resolved

cloudw commented on metaflow/plugins/kfp/kfp.py:
```diff
@@ -263,30 +264,70 @@ def _create_workflow_yaml(
             workflow["kind"] = "WorkflowTemplate"
+            workflow["spec"]["serviceAccountName"] = (
+                KUBERNETES_SERVICE_ACCOUNT or "default-editor"
+            )
```
reminder to remove 🙂

cloudw commented on metaflow/metaflow_config.py:
```diff
@@ -154,6 +154,7 @@ def from_conf(name, default=None):
 ARGO_RUN_URL_PREFIX = from_conf("ARGO_RUN_URL_PREFIX", "")
 METAFLOW_RUN_URL_PREFIX = from_conf("METAFLOW_RUN_URL_PREFIX", "")
 KFP_MAX_PARALLELISM = int(from_conf("KFP_MAX_PARALLELISM", 10))
+KFP_MAX_CONCURRENCY = int(from_conf("KFP_MAX_CONCURRENCY", 10))
```
Since this is a new parameter, do we want to make it not KFP specific and more explicitly related to the cron workflow? Maybe `CRON_MAX_CONCURRENCY`?

talebzeghmi replied:
These KFP_ prefixes are a way to namespace Metaflow config by the `KFP_` plugin, so KFP_MAX_CONCURRENCY is a sister to KFP_MAX_PARALLELISM

cloudw commented on metaflow/plugins/kfp/kfp.py:
```diff
@@ -263,30 +264,70 @@ def _create_workflow_yaml(
+            "spec": {
+                "suspend": not recurring_run_enable,
+                "schedule": (
+                    schedule if schedule else "* * 0 * *"
```
Do we want to keep using 6 digit cron?

talebzeghmi replied:
it's a five place cron (which excludes seconds):
see
* https://argoproj.github.io/argo-workflows/cron-workflows/
* https://pkg.go.dev/github.com/robfig/cron#hdr-CRON_Expression_Format

go cron allows 6 space (seconds).

This brings up a good call, to not allow an interval less than 15 minutes! created https://zbrt.atl.zillow.net/browse/AIP-7375 _argo->kfp: Limit recurring run to > 15 mins_ to be done in the webhook as a policy

cloudw commented on metaflow/plugins/kfp/kfp_cli.py:
```diff
@@ -451,6 +452,37 @@ def _convert_value(param: parameters.Parameter):
+@click.option(
+    "--recurring-run-enable/--no-recurring-run-enable",
+    "recurring_run_enable",
+    default=False,
```
For customers defaulting to `True` make more sense.

talebzeghmi replied:
It'd be nice to hear more why, shouldn't recurring runs be disabled by default given that they could use up resources and make it an explicit parameter/step for the user to take to enable recurring runs?

cloudw replied:
When user doesn't explicitly add a recurring run schedule, could you please explain what makes a suspended recurring run better than not adding recurring run?

talebzeghmi replied:
Imagine the following scenario in CICD:

**step1:**
```
python flow.py kfp create --yaml-only --pipeline-path ~/tmp/test.yaml --recurring-run-enable --recurring-run-cron "*/1 * * * *" 
```
followed by:


**step2:**
```
python flow.py kfp create --yaml-only --pipeline-path ~/tmp/test.yaml 
```

In **step2** CICD would not not about the `CronWorkflow` created in **step1**, and so it is safer to always create the CronWorkflow and mark it as disabled.

The same scenario could occur interactively through the command line (by the user via Notebook).

cloudw replied:
Is this a design for multiple flow per repo support then? I still feel it's cleaner to not add recurring run when it's needed. Maybe knowing how the user experience on CICD side looks like will help inform this decision. Do you have a design for CICD experience?

talebzeghmi replied:
https://docs.google.com/document/d/1cR-R3mqrAugM5veiceIbrwllhupTmLS8jBgBJOIuz0w/edit?pli=1#heading=h.x8ue7bngf6ku

thanks!

cloudw commented on metaflow/plugins/kfp/kfp_cli.py:
```diff
@@ -451,6 +452,37 @@ def _convert_value(param: parameters.Parameter):
+)
+@click.option(
+    "--recurring-run-concurrency",
+    "recurring_run_concurrency",
```
`recurring_run_policy` sounds more straight forward to me.

cloudw reviewed (changes requested):
How do we expect flow parameter to be passed in for Cron Workflow? Would it be part of Kustomize?

aaron-arellano commented on metaflow/metaflow_config.py:
```diff
@@ -154,6 +154,7 @@ def from_conf(name, default=None):
 ARGO_RUN_URL_PREFIX = from_conf("ARGO_RUN_URL_PREFIX", "")
 METAFLOW_RUN_URL_PREFIX = from_conf("METAFLOW_RUN_URL_PREFIX", "")
 KFP_MAX_PARALLELISM = int(from_conf("KFP_MAX_PARALLELISM", 10))
```
I know you mentioned in the meeting but feel we should not have `KFP` for these values.

talebzeghmi replied:
this is for the KFP plugin, so keeping it consistent.
In a future PR after this is all merged into `feature/kfp-argo` with https://zbrt.atl.zillow.net/browse/AIP-7385 _WFSDK: Rename KFP plugin to AIP_ added as a comment to the code.

talebzeghmi committed:
PR Feedback

talebzeghmi committed:
CronWorkflow parameters

talebzeghmi commented:
> How do we expect flow parameter to be passed in for Cron Workflow? Would it be part of Kustomize?
@cloudw I added support in this commit https://github.com/zillow/metaflow/pull/240/commits/ff719b0065cfa9cdca0b2478004c645ed65d3985

talebzeghmi committed:
Merge fixes

talebzeghmi committed:
integration test fixes

talebzeghmi committed:
only print `argo watch` if `argo` is in the path

n3o-Bhushan commented on metaflow/plugins/kfp/argo_client.py:
```diff
@@ -111,6 +111,82 @@ def register_workflow_template(self, name, workflow_template):
+                )
+
+    def create_cron_workflow(self, name: str, cron_workflow: Dict[str, Any]):
+        # Unfortunately, Kubernetes client does not handle optimistic
```
What do you mean by that? APIs should do what kubectl does right?

talebzeghmi replied:
This is actually a comment from the original argo_client.py here https://github.com/Netflix/metaflow/blob/master/metaflow/plugins/argo/argo_client.py#L38-L39

However, it seems that `kubectl` does https://en.wikipedia.org/wiki/Optimistic_concurrency_control  which is similar to https://en.wikipedia.org/wiki/Double-checked_locking# while the k8s client doesn't offer such safety

talebzeghmi replied:
sorry didn't see this till now!

n3o-Bhushan replied:
No worries, thanks! LGTM!

talebzeghmi committed:
--pipeline-name parameter

talebzeghmi committed:
KFP_ARGO_COMPILE_WORKFLOW_ONLY feature (#241)

talebzeghmi committed:
Merge branch 'feature/kfp-argo' into tz/AIP-7339-cron

talebzeghmi committed:
Merge branch 'feature/kfp-argo' into tz/AIP-7339-cron

talebzeghmi committed:
AIP-7535 cleanup creating Argo YAML (#250)

* AIP-7535 cleanup creating Argo YAML

* PR Feedback

talebzeghmi committed:
AIP-7540 Argo Parameters (#249)

* AIP-7540 Argo Parameters

* Fix for JSON parameters

* Fix exit_handler Argo failure with no Flow params

* exit handler workflow.parameters quotes

* PR Feedback

cloudw commented on metaflow/plugins/kfp/kfp.py:
```diff
@@ -193,16 +194,51 @@ def __init__(
             )
 
-            return ArgoClient(
+            cron_workflow = ArgoClient(
```
I assume this cron_workflow is no longer of `Dict[str, Any]` type. Could you please verify? If it's not maybe type hints need to be updated and this `cron_workflow` should be slightly renamed. Though I can be wrong.

talebzeghmi replied:
thanks! fixed and simplified by only returning `k8s_workflow` as the latter two parts of the return tuple were never used.

cloudw commented on metaflow/plugins/kfp/kfp.py:
```diff
@@ -242,80 +266,170 @@ def trigger(cls, kubernetes_namespace: str, name: str, parameters=None):
             # self.name is typically CamelCase as it's python class name.
             # generateName contains a sanitized version of self.name from kfp.compiler
-            workflow["metadata"]["name"] = (
+            workflow_name = (
```
[trivial] `workflow_template_name` is less ambiguous.

cloudw commented on metaflow/plugins/kfp/kfp.py:
```diff
@@ -242,80 +266,170 @@ def trigger(cls, kubernetes_namespace: str, name: str, parameters=None):
-    def create_run_on_argo(
-        self, kubernetes_namespace: str, flow_parameters: dict
+    @staticmethod
+    def _config_map(workflow_name: str, max_run_concurrency: int):
```
It looks like `max_run_concurrency` is an `Optional[int]` elsewhere. This function needs to guard against `None` value, or we drop the `Optional` in all other places.

talebzeghmi replied:
fixed!

cloudw commented on metaflow/plugins/kfp/kfp.py:
```diff
@@ -1162,6 +1295,7 @@ def _create_metaflow_step_op(
         flow_variables: FlowVariables,
         metaflow_configs: Dict[str, str],
+        flow_parameters: Dict,
         flow_parameters_json: str,
```
It looks like `flow_parameters_json` is always set to `"{{workflow.parameters}}"` whenever `_create_metaflow_step_op` function is called. Can we remove `flow_parameters_json` from parameter list of this function and use it as a string literal?

talebzeghmi replied:
nice, fixed with https://github.com/zillow/metaflow/blob/199c3320fa4c940b308fd9267e89395709f5952b/metaflow/plugins/kfp/kfp.py#L92

cloudw reviewed (commented):
LGTM. Some minor suggestions.

talebzeghmi committed:
PR Feedback
- inline {{workflow.parameters}} as FLOW_PARAMETERS_JSON
- cleanup kfp.deploy()
